servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/OptionalSslNegotiator.java

```diff
@@ -0,0 +1,248 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.netty;
+
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.http.api.HttpExecutionContext;
+import io.servicetalk.http.api.HttpLifecycleObserver;
+import io.servicetalk.http.api.HttpServerContext;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.logging.api.UserDataLoggerConfig;
+import io.servicetalk.tcp.netty.internal.ReadOnlyTcpServerConfig;
+import io.servicetalk.tcp.netty.internal.TcpServerBinder;
+import io.servicetalk.tcp.netty.internal.TcpServerChannelInitializer;
+import io.servicetalk.transport.api.ConnectionObserver;
+import io.servicetalk.transport.api.EarlyConnectionAcceptor;
+import io.servicetalk.transport.api.LateConnectionAcceptor;
+import io.servicetalk.transport.api.ServerSslConfig;
+import io.servicetalk.transport.api.SslListenMode;
+import io.servicetalk.transport.api.TransportObserver;
+import io.servicetalk.transport.netty.internal.FlushStrategy;
+import io.servicetalk.transport.netty.internal.InfluencerConnectionAcceptor;
+import io.servicetalk.transport.netty.internal.NettyConnectionContext;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelOption;
+import io.netty.handler.ssl.SslContext;
+import io.netty.util.Mapping;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.SocketAddress;
+import java.time.Duration;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import javax.annotation.Nullable;
+
+/**
+ * Negotiates SSL/non-SSL connections when SSL is enabled and {@link SslListenMode} is set to
+ * {@link SslListenMode#SSL_OPTIONAL}.
+ */
+final class OptionalSslNegotiator {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(OptionalSslNegotiator.class);
+
+    private OptionalSslNegotiator() {
+        // No instances
+    }
+
+    static Single<HttpServerContext> bind(final HttpExecutionContext executionContext,
+                                          final HttpServerConfig config,
+                                          final SocketAddress listenAddress,
+                                          @Nullable final InfluencerConnectionAcceptor connectionAcceptor,
+                                          final StreamingHttpService service,
+                                          final boolean drainRequestPayloadBody,
+                                          @Nullable final EarlyConnectionAcceptor earlyConnectionAcceptor,
+                                          @Nullable final LateConnectionAcceptor lateConnectionAcceptor) {
+        final ReadOnlyHttpServerConfig roConfig = config.asReadOnly();
```

Comment:
The caller of this `bind` method already has `roConfig`, consider passing that instead of creating yet another copy